modules/json/src/smithy4s/json/internals/JsonPayloadCodecCompilerImpl.scala

```diff
@@ -53,13 +53,25 @@ private[json] case class JsonPayloadCodecCompilerImpl(
   def fromSchema[A](schema: Schema[A], cache: Cache): PayloadCodec[A] = {
     val jcodec = jsoniterCodecCompiler.fromSchema(schema, cache)
     val reader: PayloadReader[A] = new JsonPayloadReader(jcodec)
-    val writer: PayloadWriter[A] = Writer.encodeBy { (value: A) =>
-      val intermediate = Blob(
-        writeToArray(value, jsoniterWriterConfig)(jcodec)
-      )
-      if (intermediate.sameBytesAs(Blob("null"))) Blob("{}")
-      else intermediate
-    }
+    val writer: PayloadWriter[A] =
+      schema.hints.get(smithy.api.HttpPayload) match {
```

Comment:
This is not the right fix for this : the json module MUST NO LONGER have any reference to any of the http-binding traits. That's the whole point of moving the `smithy.json` package to `smithy4s.json`

The problem is more that the original line changing nulls to empty-object is ill-located, and should be the responsibility of another component, probably [here](https://github.com/disneystreaming/smithy4s/blob/series/0.18/modules/http4s/src/smithy4s/http4s/internals/SimpleRestJsonCodecs.scala#L40-L46).

Reply:
Makes sense, thanks 👍

Reply:
Another problem is that the reliance of `org.http4s.EntityEncoder` creates the very problem of static content-types: 

https://github.com/disneystreaming/smithy4s/blob/c1f08838a9c3aeffc97540dddb1e3b77fe8e847e/modules/http4s-kernel/src/smithy4s/http4s/kernel/EntityEncoders.scala#L30-L38

Right now the path of transformation goes : 

`HttpMediaWriter => EntityEncoder => ResponseEncoder => REST-respecting ResponseEncoder`

So it may be that we want to disappear the `EntityEncoder` step in favour of a more direct `HttpMediaWriter => ResponseEncoder` that'd take care of dynamically setting the content-type based on whether the blob is empty or not. 

That change would ripple all the way to the `restSchemaCompiler` methods that currently take `CachedSchemaCompiler[EntityEncoder[F, *]]`

Reply:
Getting closer with this approach.. only thing now is that some of the compliance tests are failing because I removed the fallback logic for setting empty bodies to have a content-type of application-json because now we don't always want it set. Will keep looking at it tomorrow